elastic/kibana
# [RAC] Change index bootstrapping strategy

## Summary

This PR implements https://github.com/elastic/kibana/issues/108941. These changes aim to support additive mappings changes only.

Note:

The changes here do not currently reflect these two ACs:

- composed of the component templates matching the current version
- with the current version in the `_meta` object

As I don't *think* we need them yet, and I also want to speak to Felix about them.

## High level overview

(This may be helpful for those unfamiliar with the inner workings of this code)

- The index bootstrapping can be roughly split into two phases, the "index level resources" and the "namespaces level resources". 
  - The index level resources include the component templates and the ILM policy (if provided), these are **not** namespaced. This installation happens via `installIndexLevelResources()` and is invoked when `initializeIndex()` is called via the `RuleDataService`. Solutions, for example, will do this on plugin `setup`. You can think of these as the "building blocks" that will create the namespace level resources.
  - The namespace level resources are installed / updated when a namespace's writer is first accessed, and therefore initialised (please note that namespaces aren't used by any solution yet, but we default to `default`, so there is always a namespace). If there is no writer cached for that namespace, `initializeWriter()` will attempt to `installAndUpdateNamespaceLevelResources()`.
      - `installAndUpdateNamespaceLevelResources()` will take our "building blocks" from before and install / update the namespaced index template, create a concrete write index (if needed), and attempt to update the mappings of any current concrete write indices (via first simulating the index template via `simulateIndexTemplate()`, and then attempting to `put` those mappings).
      - If for any reason `installAndUpdateNamespaceLevelResources()` cannot be completed then writing is completely disabled.

## Testing notes

Putting aside the more obvious concerns (e.g. templates are installed correctly), the most important thing is that a developer accidentally performing a non-additive change does not break the overall alerting framework (as in, executor runs aren't broken).

Dev tools is likely to be your best friend here. Below are some useful snippets that you might want to check whilst testing (change `logs` to any other registration context):

- Check which alerts indices exist: `GET _cat/indices/.alerts-observability*`
- Get an index's template: `GET /_index_template/.alerts-observability.logs.alerts-default*`
- Check on aliases: `GET /_cat/aliases`
- Force a rollover: `POST .alerts-observability.logs.alerts-default/_rollover` (to check the ILM policy works as expected)
- See mappings for alerts indices: `GET .alerts-observability*/_mapping`
- Check a component template: `GET /_component_template/.alerts-observability.logs.alerts-mappings` (or any value from `composed_of` when checking the index template)

Things to check:

- Are the resources installed correctly with a blank slate? Does writing alerts documents work?

- If you make an **additive** change, for example to a component template, do things work as expected? (The index template should update, as should the mappings for any existing concrete write indices). You can introduce these types of additive changes here as an example: https://github.com/elastic/kibana/blob/master/x-pack/plugins/infra/server/services/rules/rule_data_client.ts#L31 (feel free to try other registration contexts / solutions)

- If you make a non-additive change, for example changing the `type` of a field, do things work as expected? (There should be a very clearly logged error, writing should be disabled, and executor runs should be unaffected). Log output should look roughly like this in that scenario:

![Screenshot 2021-09-28 at 14 11 59](https://user-images.githubusercontent.com/471693/135295720-7c22ac5d-f2bf-485f-8327-6ac834e98e1a.png)
